### PR TITLE
Implement GSP-134 Write Behavior Consistency

### PR DESCRIPTION
**Pull request by JinnyYi**

### CODE REVIEW

File: appender.go [hunk 1]
```diff
@@ -42,6 +42,41 @@ func TestAppender(t *testing.T, store types.Storager) {
...
+
+			o, err = ap.CreateAppend(path)
+
+			Convey("The second returning error also should be nil", func() {
+				So(err, ShouldBeNil)
+			})
+
+			Convey("The Object Mode should be appendable", func() {
+				// Append object's mode must be appendable.
+				So(o.Mode.IsAppend(), ShouldBeTrue)
```
Comment (Xuanwo) [inline appender.go:81]: Maybe we need to check the object size is `0`?

File: copier.go [hunk 2]
```diff
@@ -66,5 +69,92 @@ func TestCopier(t *testing.T, store types.Storager) {
...
+				Convey("The content should be match", func() {
+					So(buf, ShouldNotBeNil)
+					So(n, ShouldEqual, srcSize)
+					So(md5.Sum(buf.Bytes()), ShouldResemble, md5.Sum(content))
+				})
+			})
+		})
+
+		Convey("When Copy to an existing dir", func() {
+			d, ok := store.(types.Direr)
```
Comment (Xuanwo) [inline copier.go]: Service that didn't support `CreateDir` could be failed, we can check and skip here if not supported.

File: mover.go [hunk 3]
```diff
@@ -76,5 +77,100 @@ func TestMover(t *testing.T, store types.Storager) {
...
+				Convey("The content should be match", func() {
+					So(buf, ShouldNotBeNil)
+					So(n, ShouldEqual, srcSize)
+					So(md5.Sum(buf.Bytes()), ShouldResemble, md5.Sum(content))
+				})
+			})
+		})
+
+		Convey("When Move to an existing dir", func() {
+			d, ok := store.(types.Direr)
```
Comment (Xuanwo) [inline mover.go]: ditto.

File: storager.go [hunk 4]
```diff
@@ -75,11 +75,11 @@ func TestStorager(t *testing.T, store types.Storager) {
 		})
 
 		Convey("When Write a file", func() {
-			size := rand.Int63n(4 * 1024 * 1024) // Max file size is 4MB
-			r := io.LimitReader(randbytes.NewRand(), size)
+			fstSize := rand.Int63n(4 * 1024 * 1024) // Max file size is 4MB
```
Comment (Xuanwo) [inline storager.go]: How about using `firstSize` and `secondSize` to make it more readable?

File: appender.go [hunk 5]
```diff
@@ -42,6 +42,45 @@ func TestAppender(t *testing.T, store types.Storager) {
...
+				So(err, ShouldBeNil)
+			})
+
+			size := rand.Int63n(4 * 1024 * 1024) // Max file size is 4MB
+			r := io.LimitReader(randbytes.NewRand(), size)
+
+			_, err = ap.WriteAppend(o, r, size)
+			if err != nil {
+				t.Fatal(err)
+			}
```
Comment (xxchan) [inline appender.go:66]: Do we need to `CommitAppend` now?

File: appender.go [hunk 6]
```diff
@@ -42,6 +42,45 @@ func TestAppender(t *testing.T, store types.Storager) {
...
+			o, err := ap.CreateAppend(path)
+
+			defer func() {
+				err := store.Delete(path)
+				if err != nil {
+					t.Error(err)
+				}
+			}()
+
+			Convey("The first returning error should be nil", func() {
```
Comment (xxchan) [inline appender.go]: I think it should be "returned error", or simply "the first error"

File: appender.go [hunk 7]
```diff
@@ -42,6 +42,45 @@ func TestAppender(t *testing.T, store types.Storager) {
...
+			r := io.LimitReader(randbytes.NewRand(), size)
+
+			_, err = ap.WriteAppend(o, r, size)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			o, err = ap.CreateAppend(path)
+
+			Convey("The second returning error also should be nil", func() {
```
Comment (xxchan) [inline appender.go]: ditto

File: appender.go [hunk 8]
```diff
@@ -42,6 +42,45 @@ func TestAppender(t *testing.T, store types.Storager) {
...
+				So(err, ShouldBeNil)
+			})
+
+			Convey("The Object Mode should be appendable", func() {
+				// Append object's mode must be appendable.
+				So(o.Mode.IsAppend(), ShouldBeTrue)
+			})
+
+			Convey("The object append offset should be 0", func() {
+				So(o.MustGetAppendOffset(), ShouldBeZeroValue)
```
Comment (xxchan) [inline appender.go:85]: Is append offset the same as size?

File: copier.go [hunk 9]
```diff
@@ -66,5 +69,91 @@ func TestCopier(t *testing.T, store types.Storager) {
...
+			_, err = store.Write(dst, r, dstSize)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			defer func() {
+				err = store.Delete(src)
+				if err != nil {
+					t.Error(err)
+				}
```
Comment (xxchan) [inline copier.go]: This part should be moved ahead, otherwise `src` won't be deleted when `dst` has an error.

File: copier.go [hunk 10]
```diff
@@ -66,5 +69,91 @@ func TestCopier(t *testing.T, store types.Storager) {
...
+				 _, err = d.CreateDir(dst)
+				 if err != nil {
+					 t.Fatal(err)
+				 }
+
+				 defer func() {
+					 err = store.Delete(src)
+					 if err != nil {
+						 t.Error(err)
+					 }
```
Comment (xxchan) [inline copier.go]: ditto

File: mover.go [hunk 11]
```diff
@@ -76,5 +77,100 @@ func TestMover(t *testing.T, store types.Storager) {
...
+			_, err = store.Write(dst, r, dstSize)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			defer func() {
+				err = store.Delete(src)
+				if err != nil {
+					t.Error(err)
+				}
```
Comment (xxchan) [inline mover.go]: ditto

File: mover.go [hunk 12]
```diff
@@ -76,5 +77,100 @@ func TestMover(t *testing.T, store types.Storager) {
...
+				_, err = d.CreateDir(dst)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				defer func() {
+					err = store.Delete(src)
+					if err != nil {
+						t.Error(err)
+					}
```
Comment (xxchan) [inline mover.go]: ditto

File: multiparter.go [hunk 13]
```diff
@@ -22,17 +22,30 @@ func TestMultiparter(t *testing.T, store types.Storager) {
 			path := uuid.New().String()
 			o, err := m.CreateMultipart(path)
 
-			defer func() {
-				err := store.Delete(path, pairs.WithMultipartID(o.MustGetMultipartID()))
+			Convey("The first returning error should be nil", func() {
```
Comment (xxchan) [inline multiparter.go]: ditto

File: multiparter.go [hunk 14]
```diff
@@ -22,17 +22,30 @@ func TestMultiparter(t *testing.T, store types.Storager) {
...
 				if err != nil {
 					t.Error(err)
 				}
-			}()
+			}(o.MustGetMultipartID())
 
-			Convey("The error should be nil", func() {
+			o, err = m.CreateMultipart(path)
+
+			Convey("The second returning error also should be nil", func() {
```
Comment (xxchan) [inline multiparter.go]: ditto

File: storager.go [hunk 15]
```diff
@@ -88,7 +88,16 @@ func TestStorager(t *testing.T, store types.Storager) {
 				}
 			}()
 
-			Convey("The error should be nil", func() {
+			Convey("The first returning error should be nil", func() {
```
Comment (xxchan) [inline storager.go]: ditto

File: storager.go [hunk 16]
```diff
@@ -88,7 +88,16 @@ func TestStorager(t *testing.T, store types.Storager) {
...
+				So(err, ShouldBeNil)
+			})
+
+			secondSize := rand.Int63n(4 * 1024 * 1024) // Max file size is 4MB
+			content, _ := ioutil.ReadAll(io.LimitReader(randbytes.NewRand(), secondSize))
+
+			_, err = store.Write(path, bytes.NewReader(content), secondSize)
+
+			Convey("The second returning error also should be nil", func() {
```
Comment (xxchan) [inline storager.go]: ditto

File: appender.go (hunk 8 above)
Comment (JinnyYi) [reply appender.go:85]: Yes, the next append offset should match the current object length. 
Maybe I need to add the size(content-length) check.

Comment (xxchan) [reply appender.go:85]: Does this mean that append offset is redundant?

Comment (JinnyYi) [reply appender.go:85]: I mean we can retain the offset check and add the size check when `CreateAppend` with an existing appendable object.

Comment (xxchan) [reply appender.go:85]: I know what you mean here. Of course yes. 

But I was wondering whether we can remove the field if it's not needed.

Comment (Xuanwo) [reply appender.go:85]: Some services require `CommitAppend` to finish an append process, and until it was committed, the object will not have valid `content-length`. That's why we introduced [GSP-44: Add CommitAppend in Appender](https://github.com/beyondstorage/go-storage/blob/master/docs/rfcs/44-commit-append.md).

File: appender.go (hunk 5 above)
Comment (JinnyYi) [reply appender.go:66]: Yes, you're right. We need `CommitAppend` here as for dropbox `path` will be assigned when `CommitAppend`. If we miss `CommitAppend` here, there will be no association between the two append operations and the first append object will not be deleted.

File: appender.go (hunk 8 above)
Comment (xxchan) [reply appender.go:85]: got it

**Review by xxchan — APPROVED** [review]

**Pull request merged**